### PR TITLE
Add `test_smoke` test re-open

### PR DESCRIPTION
**Pull request by BON4**
**Issue https://github.com/canonical/postgresql-operator/issues/441**

This pull request is re-open of [pull request](https://github.com/canonical/postgresql-operator/pull/420) that was closed by me due to CLA check fix.

All this cases should be covered:

- Test Application. VM resources is removed after application removal. Install app, remove app (removal test), check no garbage left.
- Test Unit. Re-Installation. Remove unit with `--force --no-wait`, storage should be detached.
- Test Unit. Garbage ignorance. Charm should deploy in dirty environment with garbage storage.
- Test Application. Application should deploy in dirty environment with garbage storage from another application.

### CODE REVIEW

File: tests/integration/ha_tests/test_smoke.py [hunk 1]
```diff
@@ -0,0 +1,233 @@
...
+            charm,
+            application_name=APPLICATION_NAME,
+            num_units=1,
+            series=CHARM_SERIES,
+            storage={"pgdata": {"pool": "lxd-btrfs", "size": 8046}},
+            config={"profile": "testing"},
+        )
+
+        # Reducing the update status frequency to speed up the triggering of deferred events.
+        await ops_test.model.set_config({"update-status-hook-interval": "10s"})
```
[the remark this answers is not in the document]
Comment (BON4) [reply tests/integration/ha_tests/test_smoke.py]: Fixed

File: tests/integration/ha_tests/test_smoke.py [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+
+        # Check if storage exists after application deployed
+        logger.info("werifing is storage exists")
+        for attempt in Retrying(stop=stop_after_delay(15 * 3), wait=wait_fixed(3), reraise=True):
+            with attempt:
+                assert await is_storage_exists(ops_test, storage_id_str)
+
+        # Check that test database is not exists for new unit
+        logger.info("checking db")
+        assert not await check_db(ops_test, APPLICATION_NAME, TEST_DATABASE_RELATION_NAME)
```
[the remark this answers is not in the document]
Comment (BON4) [reply tests/integration/ha_tests/test_smoke.py]: Yep, there was a bug in create_db and check_db. Fixed.

File: tests/integration/ha_tests/test_smoke.py [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+        await ops_test.model.set_config({"update-status-hook-interval": "10s"})
+
+        logger.info("waiting for idle")
+        await ops_test.model.wait_for_idle(apps=[APPLICATION_NAME], status="active", timeout=1500)
+        assert ops_test.model.applications[APPLICATION_NAME].units[0].workload_status == "active"
+
+        logger.info("getting primary")
+        primary_name = await get_primary(
+            ops_test, ops_test.model.applications[APPLICATION_NAME].units[0].name
+        )
```
[the remark this answers is not in the document]
Comment (BON4) [reply tests/integration/ha_tests/test_smoke.py]: Fixed

File: tests/integration/ha_tests/test_smoke.py [hunk 4]
```diff
@@ -0,0 +1,233 @@
...
+    check_password_auth,
+    create_db,
+    get_any_deatached_storage,
+    is_postgresql_ready,
+    is_storage_exists,
+    remove_unit_force,
+    storage_id,
+)
+
+TEST_DATABASE_RELATION_NAME = "test_database"
```
[the remark this answers is not in the document]
Comment (BON4) [reply tests/integration/ha_tests/test_smoke.py]: Fixed

File: tests/integration/ha_tests/helpers.py [hunk 5]
```diff
@@ -862,3 +864,100 @@ async def reused_full_cluster_recovery_storage(ops_test: OpsTest, unit_name) ->
...
+    return db in query
+
+
+async def get_any_deatached_storage(ops_test: OpsTest) -> str:
+    """Returns any of the current available deatached storage."""
+    return_code, storages_list, stderr = await ops_test.juju(
+        "storage", "-m", f"{ops_test.controller_name}:{ops_test.model.info.name}", "--format=json"
+    )
+    if return_code != 0:
+        raise Exception(f"failed to get charm info with error: {stderr}")
```
[the remark this answers is not in the document]
Comment (BON4) [reply tests/integration/ha_tests/helpers.py]: Fixed

**Comment by dragomirp** [comment]
Hi, @BON4, please fix the linting errors.

**Review by dragomirp — APPROVED** [review]

**Comment by taurus-forever** [comment]
@marceloneppel can we merge this? P.S. Tests finally happy! \o/

**Review by marceloneppel — APPROVED** [review]

**Pull request merged**